packages/squiggle-lang/src/public/types.ts

```diff
@@ -1,13 +1,73 @@
+import { isBindingStatement } from "../ast/utils.js";
 import { result } from "../utility/result.js";
+import { vString } from "../value/index.js";
+import { VDict } from "../value/VDict.js";
 import { SqError } from "./SqError.js";
-import { SqValue } from "./SqValue/index.js";
+import { ProjectItemOutput } from "./SqProject/ProjectItem.js";
+import { SqValue, wrapValue } from "./SqValue/index.js";
 import { SqDict } from "./SqValue/SqDict.js";
+import { SqValueContext } from "./SqValueContext.js";
+import { SqValuePath, ValuePathRoot } from "./SqValuePath.js";
 
-export type SqOutput = {
+export class SqOutput {
   result: SqValue;
   bindings: SqDict;
   imports: SqDict;
   exports: SqDict;
-};
+  raw: ProjectItemOutput; // original output, not upgraded to SqValues - useful if you want to do serialization
+
+  private constructor(params: {
+    result: SqValue;
+    bindings: SqDict;
+    imports: SqDict;
+    exports: SqDict;
+    raw: ProjectItemOutput;
+  }) {
+    this.result = params.result;
+    this.bindings = params.bindings;
+    this.imports = params.imports;
+    this.exports = params.exports;
+    this.raw = params.raw;
+  }
+
+  static fromProjectItemOutput(output: ProjectItemOutput) {
```

Comment:
I'm quite happy that I managed to move this code from `SqProject`.

`SqProject` contains too much logic related to different aspects of its behavior, and if we move more stuff from it then it'll be easier to grok what's going on there.

Also, it's a clear responsibility boundary — this method doesn't depend on any project fields, it just upgrades the internal output to `SqOutput`.